text/0309-weak-cross-stack-references.md

```diff
@@ -0,0 +1,194 @@
+  }
+}
+```
+
+---
+
+Ticking the box below indicates that the public API of this RFC has been
+signed-off by the API bar raiser (the `api-approved` label was applied to the
+RFC pull request):
+
+```
+[ ] Signed-off by API Bar Raiser @xxxxx
+```
+
+## Public FAQ
+
+### What are we launching today?
+
+This is an addition to the core CDK library, adding more flexibility to configuration of the reference style.
+
+### Why should I use this feature?
+
+You can use this feature if the standard reference style is causing you too many problems.
+
+## Internal FAQ
+
+### Why are we doing this?
+
+It comes up frequently that users are stuck trying to remove or refactor a resource, only to find out that they can't because the resource is referenced in another stack. This takes a lot of work to disentangle. This problem is so common that it has been named the "deadly embrace" problem (mostly occurs in Pipelines, where users are unable to play with the deployment order of stacks).
+
+A better mechanism to set up references, combined with sensible defaults for resources that are not impacted by such changes like
+Dashboards, should make this problem less frequent.
+
+### Why should we _not_ do this?
+
+It might not be worth the additional complexity in the token and reference system (which is already pretty complex).
+
+### What is the technical solution (design) of this feature?
+
+Transporting the value from one stack to another: we will generate SSM Parameter Store values in the producing stack, and consume them (as `Parameters`) in the consuming stack.
+
+To configure the individual references to be either strong or weak, we:
+
+- add a `referenceStyle` field to `IResolveContext`
+- `stack.resolve()` will set the initial `referenceStyle` to the stack configuration
+- make `weakStringReference` return a new `IResolvable` that in its `resolve()` method will resolve the inner value with `referenceStyle` set to the appropriate reference style
+- references will check the `referenceStyle` of the context and call a new method on `Stack` to create the SSM parameter export and import
+
+### What alternative solutions did you consider?
+
+There are two possible alternatives:
+
+- Explicit exporting
+- Transporting values happens outside CloudFormation
+
+### Explicit exporting
+
+We could punt this concern to users more explicitly. Every (owned) construct should have a method `export()` that works like this:
+
+```ts
+
+const producer = new ProducerStack(this, 'Producer');
+
+new ConsumerStack(this, 'Consumer', {
+  bucket: producer.bucket.export(),
+});
+```
+
+The `export` method for a `Bucket` should look like this:
+
+```ts
+class Bucket {
+  public export(options?: ExportOptions): IBucket;
+}
+```
+
+And use either `Exports` or SSM parameters to reference itself.
+
+The explicitness puts users in full control, able to generate exports in either or both styles, even without a consumer.
```

Comment:
This makes it more generic when compared to the feature flag, right?

Reply:
I don't know what generic means in this context. But it sure makes it more *explicit*, at the cost of:

* having to type more
* having to be aware when you are passing constructs across stacks (might not be obvious when writing higher-level code) 
* losing the ability to make decisions on a per-consumption basis (person passing into the consumer stack makes the choice for the entire stack)
* losing the ability to change the producing construct from the consumer stack